{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04012816",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "sys.path.append(os.getcwd())\n",
    "from STdGCN.STdGCN import run_STdGCN\n",
    "\n",
    "'''\n",
    "This module is used to provide the path of the loading data and saving data.\n",
    "\n",
    "Parameters:\n",
    "sc_path: The path for loading single cell reference data.\n",
    "ST_path: The path for loading spatial transcriptomics data.\n",
    "output_path: The path for saving output files.\n",
    "\n",
    "The relevant file name and data format for loading:\n",
    "sc_data.tsv: The expression matrix of the single cell reference data with cells as rows and genes as columns. This file should be saved in \"sc_path\".\n",
    "sc_label.tsv: The cell-type annotation of sincle cell data. The table should have two columns: The cell barcode/name and the cell-type annotation information.\n",
    "            This file should be saved in \"sc_path\".\n",
    "ST_data.tsv: The expression matrix of the spatial transcriptomics data with spots as rows and genes as columns. This file should be saved in \"ST_path\".\n",
    "coordinates.csv: The coordinates of the spatial transcriptomics data. The table should have three columns: Spot barcode/name, X axis (column name 'x'), and Y axis (column name 'y').\n",
    "            This file should be saved in \"ST_path\".\n",
    "marker_genes.tsv [optional]: The gene list used to run STdGCN. Each row is a gene and no table header is permitted. This file should be saved in \"sc_path\".\n",
    "ST_ground_truth.tsv [optional]: The ground truth of ST data. The data should be transformed into the cell type proportions. This file should be saved in \"ST_path\".\n",
    "'''\n",
    "paths = {\n",
    "    'sc_path': './data/sc_data',\n",
    "    'ST_path': './data/ST_data',\n",
    "    'output_path': './output',\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "'''\n",
    "This module is used to preprocess the input data and identify marker genes [optional].\n",
    "\n",
    "Parameters:\n",
    "'preprocess': [bool]. Select whether the input expression data needs to be preprocessed. This step includes normalization, logarithmization, selecting highly variable genes, \n",
    "                    regressing out mitochondrial genes, and scaling data.\n",
    "'normalize': [bool]. When 'preprocess'=True, select whether you need to normalize each cell/spot by total counts = 10,000, so that every cell/spot has the same total \n",
    "                    count after normalization.\n",
    "'log': [bool]. When 'preprocess'=True, select whether you need to logarithmize (X=log(X+1)) the expression matrix.\n",
    "'highly_variable_genes': [bool]. When 'preprocess'=True, select whether you need to filter the highly variable genes.\n",
    "'highly_variable_gene_num': [int or None]. When 'preprocess'=True and 'highly_variable_genes'=True, select the number of highly-variable genes to keep.\n",
    "'regress_out': [bool]. When 'preprocess'=True, select whether you need to regress out mitochondrial genes.\n",
    "'scale': [bool]. When 'preprocess'=True, select whether you need to scale each gene to unit variance and zero mean.\n",
    "'PCA_components': [int]. Number of principal components to compute for principal component analysis (PCA).\n",
    "'marker_gene_method': ['logreg', 'wilcoxon']. We used \"scanpy.tl.rank_genes_groups\" (https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.rank_genes_groups.html)\n",
    "                    to identify cell type marker genes. For marker gene selection, STdGCN provides two methods, 'wilcoxon' (Wilcoxon rank-sum) and 'logreg' (uses \n",
    "                    logistic regression). \n",
    "'top_gene_per_type': [int]. The number of genes for each cell type that can be used to train STdGCN.\n",
    "'filter_wilcoxon_marker_genes': [bool]. When 'marker_gene_method'='wilcoxon', select whether you need additional steps for gene filtering.\n",
    "'pvals_adj_threshold': [float or None]. When 'marker_gene_method'='wilcoxon' and 'rank_gene_filter'=True, only genes with corrected p-values < 'pvals_adj_threshold' were kept.\n",
    "'log_fold_change_threshold': [float or None]. When 'marker_gene_method'='wilcoxon' and 'rank_gene_filter'=True, only genes with log fold change > 'log_fold_change_threshold' were kept.\n",
    "'min_within_group_fraction_threshold': [float or None]. When 'marker_gene_method'='wilcoxon' and 'rank_gene_filter'=True, only genes expressed with fraction at least\n",
    "                    'min_within_group_fraction_threshold' in the cell type were kept.\n",
    "'max_between_group_fraction_threshold': [float or None]. When 'marker_gene_method'='wilcoxon' and 'rank_gene_filter'=True, only genes expressed with fraction at most \n",
    "                    'max_between_group_fraction_threshold' in the union of the rest of cell types were kept.\n",
    "'''\n",
    "find_marker_genes_paras = {\n",
    "    'preprocess': True,\n",
    "    'normalize': True,\n",
    "    'log': True,\n",
    "    'highly_variable_genes': False,\n",
    "    'highly_variable_gene_num': None,\n",
    "    'regress_out': False,\n",
    "    'PCA_components': 30, \n",
    "    'marker_gene_method': 'logreg',\n",
    "    'top_gene_per_type': 100,\n",
    "    'filter_wilcoxon_marker_genes': True,\n",
    "    'pvals_adj_threshold': 0.10,\n",
    "    'log_fold_change_threshold': 1,\n",
    "    'min_within_group_fraction_threshold': None,\n",
    "    'max_between_group_fraction_threshold': None,\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "'''\n",
    "This module is used to simulate pseudo-spots.\n",
    "\n",
    "Parameters:\n",
    "'spot_num': [int]. The number of pseudo-spots.\n",
    "'min_cell_num_in_spot': [int]. The minimum number of cells in a pseudo-spot.\n",
    "'max_cell_num_in_spot': [int]. The maximum number of cells in a pseudo-spot.\n",
    "'generation_method': ['cell' or 'celltype']. STdGCN provides two pseudo-spot simulation methods. When 'generation_method'='cell', each cell is equally selected. When \n",
    "                    'generation_method'='celltype', each cell type is equally selected. See manuscript for more details.\n",
    "'max_cell_types_in_spot': [int]. When 'generation_method'='celltype', choose the maximum number of cell types in a pseudo-spot.\n",
    "'''\n",
    "pseudo_spot_simulation_paras = {\n",
    "    'spot_num': 30000,\n",
    "    'min_cell_num_in_spot': 8,\n",
    "    'max_cell_num_in_spot': 12,\n",
    "    'generation_method': 'celltype',\n",
    "    'max_cell_types_in_spot': 4,   \n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "'''\n",
    "This module is used for real- and pseudo- spots normalization.\n",
    "\n",
    "Parameters:\n",
    "'normalize': [bool]. Select whether you need to normalize each cell/spot by total counts = 10,000, so that every cell/spot has the same total count after normalization.\n",
    "'log': [bool]. Select whether you need to logarithmize (X=log(X+1)) the expression matrix.\n",
    "'scale': [bool]. Select whether you need to scale each gene to unit variance and zero mean.\n",
    "'''\n",
    "data_normalization_paras = {\n",
    "    'normalize': True, \n",
    "    'log': True, \n",
    "    'scale': False,\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "'''\n",
    "This module is used to integrate the normalized real- and pseudo- spots together to construct the real-to-pseudo-spot link graph.\n",
    "\n",
    "Parameters:\n",
    "'batch_removal_method': ['mnn', 'scanorama', 'combat', None]. Considering batch effects, STdGCN provides four integration methods: mnn (mnnpy, DOI:10.1038/nbt.4091),\n",
    "                    scanorama (Scanorama, DOI: 10.1038/s41587-019-0113-3), combat (Combat, DOI: 10.1093/biostatistics/kxj037), None (concatenation with no batch removal).\n",
    "'dimensionality_reduction_method': ['PCA', 'autoencoder', 'nmf', None]. When 'batch_removal_method' is not 'scanorama', select whether the data needs dimensionality reduction, and which\n",
    "                    dimensionality reduction method is applied.\n",
    "'dim': [int]. When 'batch_removal_method'='scanorama', select the dimension for this method. When 'batch_removal_method' is not 'scanorama' and 'dimensionality_reduction_method' is\n",
    "                    not None, select the dimension of the dimensionality reduction.\n",
    "'scale': [bool]. When 'batch_removal_method' is not 'scanorama', select whether you need to scale each gene to unit variance and zero mean.\n",
    "'''\n",
    "integration_for_adj_paras = {\n",
    "    'batch_removal_method': None, \n",
    "    'dim': 30, \n",
    "    'dimensionality_reduction_method': 'PCA',\n",
    "    'scale': True,\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "'''\n",
    "The module is used to construct the adjacency matrix of the expression graph, which contains three subgraphs: a real-to-pseudo-spot graph, a pseudo-spots internal graph, \n",
    "and a real-spots internal graph.\n",
    "\n",
    "Parameters:\n",
    "'find_neighbor_method' ['MNN', 'KNN']. STdGCN provides two methods for link graph construction, KNN (K-nearest neighbors) and MNN (mutual nearest neighbors, DOI: 10.1038/nbt.4091).\n",
    "'dist_method': ['euclidean', 'cosine']. The metrics used for computing paired distances between spots.\n",
    "'corr_dist_neighbors': [int]. The number of nearest neighbors.\n",
    "'PCA_dimensionality_reduction': [bool]. For pseudo-spots internal graph and real-spots internal graph construction, select if the data needs to use PCA dimensionality reduction before\n",
    "                    computing paired distances between spots.\n",
    "'dim': [int]. When 'PCA_dimensionality_reduction'=True, select the dimension of the PCA.\n",
    "'''\n",
    "inter_exp_adj_paras = {\n",
    "    'find_neighbor_method': 'MNN', \n",
    "    'dist_method': 'cosine', \n",
    "    'corr_dist_neighbors': 20, \n",
    "}\n",
    "real_intra_exp_adj_paras = {\n",
    "    'find_neighbor_method': 'MNN', \n",
    "    'dist_method': 'cosine',  \n",
    "    'corr_dist_neighbors': 10,\n",
    "    'PCA_dimensionality_reduction': False,\n",
    "    'dim': 50,\n",
    "}\n",
    "pseudo_intra_exp_adj_paras = {\n",
    "    'find_neighbor_method': 'MNN', \n",
    "    'dist_method': 'cosine', \n",
    "    'corr_dist_neighbors': 20,\n",
    "    'PCA_dimensionality_reduction': False,\n",
    "    'dim': 50,\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "'''\n",
    "The module is used to construct the adjacency matrix of the spatial graph.\n",
    "\n",
    "Parameters:\n",
    "'space_dist_threshold': [float or None]. Only the distance between two spots smaller than 'space_dist_threshold' can be linked.\n",
    "'link_method' ['soft', 'hard']. If spot i and j linked, A(i,j)=1 if 'link_method'='hard', while A(i,j)=1/distance(i,j) if 'link_method'='soft'. See manuscript for more details.\n",
    "'''\n",
    "spatial_adj_paras = {\n",
    "    'link_method': 'soft', \n",
    "    'space_dist_threshold': 2,\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "'''\n",
    "This module is used to integrate the normalized real- and pseudo- spots as the input feature for STdGCN.\n",
    "\n",
    "Parameters:\n",
    "'batch_removal_method': ['mnn', 'scanorama', 'combat', None]. Considering batch effects, STdGCN provides four integration methods: mnn (mnnpy, DOI:10.1038/nbt.4091),\n",
    "                    scanorama (Scanorama, DOI: 10.1038/s41587-019-0113-3), combat (Combat, DOI: 10.1093/biostatistics/kxj037), None (concatenation with no batch removal).\n",
    "'dimensionality_reduction_method': ['PCA', 'autoencoder', 'nmf', None]. When 'batch_removal_method' is not 'scanorama', select whether the data needs dimensionality reduction, and which\n",
    "                    dimensionality reduction method is applied.\n",
    "'dim': [int]. When 'batch_removal_method'='scanorama', select the dimension for this method. When 'batch_removal_method' is not 'scanorama' and 'dimensionality_reduction_method' is\n",
    "                    not None, select the dimension of the dimensionality reduction.\n",
    "'scale': [bool]. When 'batch_removal_method' is not 'scanorama', select whether you need to scale each gene to unit variance and zero mean.\n",
    "'''\n",
    "integration_for_feature_paras = {\n",
    "    'batch_removal_method': None, \n",
    "    'dimensionality_reduction_method': None, \n",
    "    'dim': 80,\n",
    "    'scale': True,\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "'''\n",
    "This module is used for setting the deep learning parameters for STdGCN.\n",
    "\n",
    "Parameters:\n",
    "'epoch_n': [int]. The maximum number of epochs.\n",
    "'dim': [int]. The dimension of the hidden layers.\n",
    "'common_hid_layers_num': [int]. The number of GCN layers = 'common_hid_layers_num'+1.\n",
    "'fcnn_hid_layers_num': [int]. The number of fully connected neural network layers = 'fcnn_hid_layers_num'+2.\n",
    "'dropout': [float]. The probability of an element to be zeroed.\n",
    "'learning_rate_SGD': [float]. Initial learning rate.\n",
    "'weight_decay_SGD': [float]. L2 penalty.\n",
    "'momentum': [float]. Momentum factor.\n",
    "'dampening': [float]. Dampening for momentum.\n",
    "'nesterov': [bool]. Enables Nesterov momentum.\n",
    "'early_stopping_patience': [int]. Early stopping epochs.\n",
    "'clip_grad_max_norm': [float]. Clips gradient norm of an iterable of parameters.\n",
    "'LambdaLR_scheduler_coefficient': [float]. The coefficent of the LambdaLR scheduler fucntion:  lr(epoch) = [LambdaLR_scheduler_coefficient] ^ epoch_n Ã— learning_rate_SGD. \n",
    "'print_loss_epoch_step': [int]. Print the loss value at every 'print_epoch_step' epoch.\n",
    "'cpu_num': [int]. Set the number of threads used for intraop parallelism on CPU.\n",
    "'''\n",
    "GCN_paras = {\n",
    "    'epoch_n': 3000,\n",
    "    'dim': 80,\n",
    "    'common_hid_layers_num': 1,\n",
    "    'fcnn_hid_layers_num': 1,\n",
    "    'dropout': 0,\n",
    "    'learning_rate_SGD': 5e-1,\n",
    "    'weight_decay_SGD': 3e-4,\n",
    "    'momentum': 0.9,\n",
    "    'dampening': 0,\n",
    "    'nesterov': True,\n",
    "    'early_stopping_patience': 20,\n",
    "    'clip_grad_max_norm': 1,\n",
    "    'LambdaLR_scheduler_coefficient': 0.997,\n",
    "    'print_loss_epoch_step': 10,\n",
    "    'cpu_num': 10,\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "'''\n",
    "## run STdGCN\n",
    "\n",
    "Parameters\n",
    "'load_test_groundtruth': [bool]. Select whether you need to upload the ground truth file (ST_ground_truth.tsv) of the spatial transcriptomics data to track the performance of STdGCN.\n",
    "'use_marker_genes': [bool]. Select whether you need the gene selection process before running STdGCN. Otherwise use common genes from single cell and spatial transcriptomics data.\n",
    "'external_genes': [bool]. When \"use_marker_genes\"=True, you can upload your specified gene list (marker_genes.tsv) to run STdGCN.\n",
    "'generate_new_pseudo_spots': [bool]. STdGCN will save the simulated pseudo-spots to \"pseudo_ST.pkl\". If you want to run multiple deconvolutions with the same single cell reference data,\n",
    "                    you don't need to simulate new pseudo-spots and set 'generate_new_pseudo_spots'=False. When 'generate_new_pseudo_spots'=False, you need to pre-move the \"pseudo_ST.pkl\" \n",
    "                    to the 'output_path' so that STdGCN can directly load the pre-simulated pseudo-spots.\n",
    "'fraction_pie_plot': [bool]. Select whether you need to draw the pie plot of the predicted results. Based on our experience, we do not recommend to draw the pie plot when the predicted\n",
    "                    spot number is very large. For 1,000 spots, the plotting time is less than 3 minutes; for 2,000 spots, the plotting time is about 20 minutes; for 3,000 spots, it takes\n",
    "                    about one hour.\n",
    "'cell_type_distribution_plot': [bool]. Select whether you need to draw the scatter plot of the predicted results for each cell type.\n",
    "'''\n",
    "results =  run_STdGCN(paths,\n",
    "                          load_test_groundtruth = True,\n",
    "                          use_marker_genes = True,\n",
    "                          external_genes = False,\n",
    "                          find_marker_genes_paras = find_marker_genes_paras,\n",
    "                          generate_new_pseudo_spots = False, \n",
    "                          pseudo_spot_simulation_paras = pseudo_spot_simulation_paras,\n",
    "                          data_normalization_paras = data_normalization_paras,\n",
    "                          integration_for_adj_paras = integration_for_adj_paras,\n",
    "                          inter_exp_adj_paras = inter_exp_adj_paras,\n",
    "                          spatial_adj_paras = spatial_adj_paras,\n",
    "                          real_intra_exp_adj_paras = real_intra_exp_adj_paras,\n",
    "                          pseudo_intra_exp_adj_paras = pseudo_intra_exp_adj_paras,\n",
    "                          integration_for_feature_paras = integration_for_feature_paras,\n",
    "                          GCN_paras = GCN_paras,\n",
    "                          fraction_pie_plot = True,\n",
    "                          cell_type_distribution_plot = True\n",
    "                         )\n",
    "\n",
    "results.write_h5ad(paths['output_path']+'/results.h5ad')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
